tests/k8st/tests/test_bgp_advert.py

```diff
@@ -211,109 +176,31 @@ def add_svc_external_ips(self, svc, ns, ips):
 
-    def test_rr(self):
-        self.tearDown()
-        self.setUpRR()
-
-        # Create ExternalTrafficPolicy Local service with one endpoint on node-1
-        kubectl("""apply -f - << EOF
-apiVersion: apps/v1
-kind: Deployment
-metadata:
-  name: nginx-rr
-  namespace: bgp-test
-  labels:
-    app: nginx
-spec:
-  replicas: 1
-  selector:
-    matchLabels:
-      app: nginx
-      run: nginx-rr
-  template:
-    metadata:
-      labels:
-        app: nginx
-        run: nginx-rr
-    spec:
-      containers:
-      - name: nginx-rr
-        image: nginx:1.7.9
-        ports:
-        - containerPort: 80
-      nodeSelector:
-        beta.kubernetes.io/os: linux
-        kubernetes.io/hostname: %s
----
-apiVersion: v1
-kind: Service
-metadata:
-  name: nginx-rr
-  namespace: bgp-test
-  labels:
-    app: nginx
-    run: nginx-rr
-spec:
-  externalIPs:
-  - 175.200.1.1
-  ports:
-  - port: 80
-    targetPort: 80
-  selector:
-    app: nginx
-    run: nginx-rr
-  type: NodePort
-  externalTrafficPolicy: Local
-EOF
-""" % self.nodes[1])
 
-        calicoctl("get nodes -o yaml")
-        calicoctl("get bgppeers -o yaml")
-        calicoctl("get bgpconfigs -o yaml")
+class TestBGPAdvert(_TestBGPAdvert):
 
-        # Update the node-2 to behave as a route-reflector
-        json_str = calicoctl("get node %s -o json" % self.nodes[2])
-        node_dict = json.loads(json_str)
-        node_dict['metadata']['labels']['i-am-a-route-reflector'] = 'true'
-        node_dict['spec']['bgp']['routeReflectorClusterID'] = '224.0.0.1'
-        calicoctl("""apply -f - << EOF
-%s
-EOF
-""" % json.dumps(node_dict))
+    # In the tests of this class we have a full BGP mesh between the
+    # cluster nodes (kube-control-plane, kube-node-1 and kube-node-2)
+    # and the external node (kube-node-extra):
+    #
+    # - The full mesh between the cluster nodes is configured by
+    #   nodeToNodeMeshEnabled: true.
+    #
+    # - The peerings from each cluster node to the external node are
+    #   configured by self.get_extra_peer_spec().
+    #
+    # - The peerings from the external node to each cluster node are
+    #   configured by self.get_bird_conf().
 
-        # Disable node-to-node mesh and add cluster and external IPs CIDRs to advertise.
-        # Configure bgp peering between node-1 and RR and also between external node and RR.
-        calicoctl("""apply -f - << EOF
-apiVersion: projectcalico.org/v3
-kind: BGPConfiguration
-metadata:
-  name: default
-spec:
-  nodeToNodeMeshEnabled: false
-  asNumber: 64512
-  serviceClusterIPs:
-  - cidr: 10.96.0.0/12
-  serviceExternalIPs:
-  - cidr: 175.200.0.0/16
-EOF
-""")
+    def get_bird_conf(self):
+        return bird_conf % (self.ips[0], self.ips[1], self.ips[2], self.ips[3])
 
-        calicoctl("""apply -f - << EOF
-apiVersion: projectcalico.org/v3
-kind: BGPPeer
-metadata: {name: kube-node-1}
+    def get_extra_peer_spec(self):
+        return """
```

Comment:
What this is for?

Reply:
It specifies the BGPPeer for the peering(s) to the external (i.e. non-cluster) node, and allows that to be different for the two test classes, i.e. the non-RR and the RR class.
- In the non-RR class, **all** the cluster nodes peer with the external node.
- In the RR class, the cluster nodes peer with the in-cluster RR node, and only the RR node peers with the external node.

Does that answer your question?

Reply:
I see. Sorry I mistakenly thought all the functions just return "" but actually they are not.